src/components/StudyEnd.tsx

```diff
@@ -100,12 +104,42 @@ export function StudyEnd() {
     checkDataCollectionEnabled();
   }, [storageEngine, studyId]);
 
+  const processedStudyEndMsg = useMemo(() => {
+    const { studyEndMsg } = studyConfig.uiConfig;
+    const { urlParticipantIdParam } = studyConfig.uiConfig;
+
+    if (!urlParticipantIdParam || !studyEndMsg?.includes('{')) {
+      return studyEndMsg;
+    }
+
+    const templateData: Record<string, string> = { PARTICIPANT_ID: participantId };
+
+    const responses = Object.values(studyConfig.components)
+      .flatMap((component) => (
+        Array.isArray((component as { response?: Response[] }).response)
+          ? (component as { response: Response[] }).response
+          : []
+      ));
```

Comment:
The reason this was likely complaining about types is because of base components. Using the type you defined here adds a bug with base components, because responses are not always guaranteed to be available at the `components` level (e.g. they are defined on the base component instead). You will need to use the function that takes a component and fully specifies it. That will also let you remove this type cast.